### ISSUE TITLE
TypeScript - Support for Readonly vectors and matrices

### ISSUE THREAD

**Issue by AlexMax** [toji/gl-matrix#381]
The new included type file is convenient, but there is a shortcoming when it comes to my workflow.  It appears you cannot pass `Readonly` vectors or matrices to any function.  Here's an example:

```typescript
const a: Readonly<vec2> = vec2.fromValues(1, 2);
const b = vec2.clone(a);
```

This gives the following error:

```
Argument of type 'readonly [number, number] | Readonly<Float32Array>' is not assignable to parameter of type 'vec2'.
  Type 'readonly [number, number]' is not assignable to type 'vec2'.
    The type 'readonly [number, number]' is 'readonly' and cannot be assigned to the mutable type '[number, number]'.ts(2345)
```

So I thought I would ask if the project would consider adding support for `Readonly` arrays in its typing file  This change would involve going through each function, seeing if the function mutates the parameter, and if not, changing the type from `vec2` to `Readonly<vec2>` and so on.  Input paramters would be made readonly, but variables like `out` would not.

**This is work I'm willing to do**, as I used to have an in-tree typings file with proper readonly annotations before the update broke it.  And instead of trying to fix up my build to get it working again, I thought I could do more people a favor than just myself.

That said, I know that not everybody uses Readonly in their own code, and it's worth mentioning that marking a parameter readonly does not prevent you from passing a non-readonly vector.

```typescript
function testing(i: Readonly<vec2>) {
    console.debug(i);
}
testing(vec2.fromValues(1, 2));
```

That still works - users of the library do not have to update their code unless they want to.

Also, if you want, we can be a little less specific (and perhaps more flexible for future possibilities) with something like:

```typescript
type Input<T> = Readonly<T>;

// meanwhile, elsewhere...

export function copy(out: vec2, a: Input<vec2>): vec2;
```
That way, if we ever need to augment the Input type later on down the line we can just update `Input<T>` instead of changing every parameter again.

So, to summarize is this a feature the project would be interested in?

**Comment by stefnotch** [comment]
That is a really cool suggestion! Yes, a PR would be rather appreciated. I'd wait a few days before doing so though, just to give other people some time to comment.

If I didn't overlook anything, the only parameter that should ever be mutated is the `out` parameter. Everything else is probably a bug.

**Comment by AlexMax** [comment]
Okay, so something I figured out during implementation is that using `Readonly<vec2>` is a JSDoc does not seem to agree with typescript's d.ts generator.  It gets converted into `readonly [number, number] | Readonly<Float32Array>` which is rather noisy.

However, I was able to sidestep the issue by creating custom types.

```typescript
export type ReadonlyVec2 = readonly [number, number] | Float32Array;
```

I was able to apply this parameter type change project-wide with a regex that skipped `out` parameters, and thus far I've been able to use the resulting library in my own codebase to good effect - though I honestly would rather write a couple of typescript-specific tests on the gl-matrix end just to be sure.

It's worth nothing that `Readonly` is not sound when applied to typed arrays like `Float32Array`, and in the absence of any official `ReadonlyFloat32Array` I have left it as the original version.  This is perfectly fine, though, because of the way types work.

```typescript
const f: ReadonlyVec2 = vec2.create();
console.debug(f);
f.fill(2);
console.debug(f);
```

This gives you an error, regardless of what the underlying type of `f` is:

```

**Comment by AlexMax** [comment]
Sorry about the issue close - I had left a browser window open with a half-finished comment on it, and I accidentally hit something that made it submit and close.  This is my best attempt at hurriedly finishing the comment, sorry if some things don't make sense.

- - -

Okay, so something I figured out during implementation is that using `Readonly<vec2>` is a JSDoc does not seem to agree with typescript's d.ts generator.  It gets converted into `readonly [number, number] | Readonly<Float32Array>` which is rather noisy.

However, I was able to sidestep the issue by creating custom types.

```typescript
export type ReadonlyVec2 = readonly [number, number] | Float32Array;
```

I was able to apply this parameter type change project-wide with a regex that skipped `out` parameters, and thus far I've been able to use the resulting library in my own codebase to good effect - though I honestly would rather write a couple of typescript-specific tests on the gl-matrix end just to be sure.

It's worth nothing that `Readonly` is not sound when applied to typed arrays like `Float32Array`, and in the absence of any official `ReadonlyFloat32Array` I have left it as the original version.  This is perfectly fine, though, because of the way types work.

```typescript
const f: ReadonlyVec2 = vec2.create();
console.debug(f);
f.fill(2);
console.debug(f);
```

This gives you an error, regardless of what the underlying type of `f` is:

```
Property 'fill' does not exist on type 'ReadonlyVec2'.
  Property 'fill' does not exist on type 'readonly [number, number]'.(2339)
```

**Comment by stefnotch** [comment]
Thank you for investigating!

**Comment by AlexMax** [comment]
I went ahead and pushed what I have to a branch in a fork.

https://github.com/AlexMax/gl-matrix/tree/readonly-types

I have not written tests yet, because I'm unsure what the appropriate strategy is here.  For one, the testing apparatus does not support TypeScript without changes, and for another, any functionality I would want to test would not be of runtime behavior, but of compiler behavior - ensuring that certain constructs compile and certain other constructs properly fail to compile.

In any case, feel free to clone that branch and build the docs and dist if you want to take a look at it - I don't plan on commiting dist/docs changes until the branch is "done".

**Comment by AlexMax** [comment]
@stefnotch have you had an opportunity to look at the branch yet?

I've been a little busy this week, but I've been rolling around in my head the idea of having tests for the types and I'm still not sure what the best play is.  The trouble is, whereas your existing tests are running against the original source files, the only way I can picture testing with types working is against the distribution, where you actually have access to the final `.d.ts` file.

I guess there's no clear obvious path forward here.  To be frank, the things I want to test are kind of trivial, stuff like "does this type actually exist, can I pass this type to this parameter, do i get an error if I pass this type to this other kind of parameter".  It's not like there were tests for the types before, so maybe they're not even necessary.

So I guess I'll just defer to you (or another maintainer) - what are expectations and requests before you would like me to open a pull request?

**Comment by stefnotch** [comment]
I think it'd be fine without any tests for now. 

And yes, the best way to test the `.d.ts` file would probably be to check if it's valid and whatnot right after generating it.

**Comment by AlexMax** [comment]
Okay I updated the branch to type-check the bundle after it's assembled.  Would you like me to open a pull request at this point?  Do I need to commit the build artifacts at this point, or is that you you-thing?

**Comment by stefnotch** [comment]
Looks good, please do open a PR.
Regarding the build artifacts, I'll take care of them.
